pkg/server/sotw/v3/xds.go

```diff
@@ -0,0 +1,163 @@
+	"github.com/envoyproxy/go-control-plane/pkg/server/stream/v3"
+)
+
+// process handles a bi-di stream request
+func (s *server) process(str stream.Stream, reqCh chan *discovery.DiscoveryRequest, defaultTypeURL string) error {
+	// create our streamWrapper which can be passed down to sub control loops.
+	// this is useful for abstracting critical information for various types of
+	// xDS resource processing.
+	sw := streamWrapper{
+		stream:    str,
+		ID:        atomic.AddInt64(&s.streamCount, 1), // increment stream count
+		callbacks: s.callbacks,
+		node:      &core.Node{}, // node may only be set on the first discovery request
+
+		// a collection of stack allocated watches per request type.
+		watches:                newWatches(),
+		streamState:            stream.NewStreamState(false, map[string]string{}),
+		lastDiscoveryResponses: make(map[string]lastDiscoveryResponse),
+	}
+
+	// cleanup once our stream has ended.
+	defer sw.shutdown()
+
+	if s.callbacks != nil {
+		if err := s.callbacks.OnStreamOpen(str.Context(), sw.ID, defaultTypeURL); err != nil {
+			return err
+		}
+	}
+
+	// do an initial recompute so we can load the first 2 channels:
+	// <-reqCh
+	// s.ctx.Done()
+	sw.watches.recompute(s.ctx, reqCh)
+
+	for {
+		// The list of select cases looks like this:
+		// 0: <- ctx.Done
+		// 1: <- reqCh
+		// 2...: per type watches
+		index, value, ok := reflect.Select(sw.watches.cases)
+		switch index {
+		// ctx.Done() -> if we receive a value here we return
+		// as no further computation is needed
+		case 0:
+			return nil
+		// Case 1 handles any request inbound on the stream
+		// and handles all initialization as needed
+		case 1:
+			// input stream ended or failed
+			if !ok {
+				return nil
+			}
+
+			req := value.Interface().(*discovery.DiscoveryRequest)
+			if req == nil {
+				return status.Errorf(codes.Unavailable, "empty request")
+			}
+
+			// Only first request is guaranteed to hold node info so if it's missing, reassign.
+			if req.Node != nil {
+				sw.node = req.Node
+			} else {
+				req.Node = sw.node
+			}
+
+			// nonces can be reused across streams; we verify nonce only if nonce is not initialized
+			nonce := req.GetResponseNonce()
+
+			// type URL is required for ADS but is implicit for xDS
+			if defaultTypeURL == resource.AnyType {
+				if req.TypeUrl == "" {
+					return status.Errorf(codes.InvalidArgument, "type URL is required for ADS")
+				}
+
+				// When using ADS we need to order responses.
+				// This is guaranteed in the xDS protocol specification
+				// as ADS is required to be eventually consistent.
+				// More details can be found here if interested:
+				// https://www.envoyproxy.io/docs/envoy/latest/api-docs/xds_protocol#eventual-consistency-considerations
+				if s.opts.Ordered {
+					// send our first request on the stream again so it doesn't get
+					// lost in processing on the new control loop
+					go func() {
+						reqCh <- req
```

Comment:
Yea I did think about that. I couldn't really come up with anything smarter than this since we need to shuttle the request down into the processors, would be open to suggestions! If this does cause a bug we can patch after the fact too

Reply:
For now I think adding a comment about the risk might be useful.
Longer term I'd like to do more work on improving the state machine and aligning between delta and sotw, and that work will hopefully cover this issue

Reply:
Good call I can add a quick comment then we can get this merged